signature/jws/jws.go

```diff
@@ -0,0 +1,254 @@
+			delete(mustMarkedCrit, val)
+		} else {
+			if _, ok := protectedHeader.ExtendedAttributes[val]; !ok {
+				return &signature.InvalidSignatureError{
+					Msg: fmt.Sprintf("%q header is marked critical but not present", val)}
+			}
+		}
+	}
+
+	// validate all required critical headers are present.
+	if len(mustMarkedCrit) != 0 {
+		// This is not taken care by VerifySignerInfo method
+		keys := make([]string, 0, len(mustMarkedCrit))
+		for k := range mustMarkedCrit {
+			keys = append(keys, k)
+		}
+		return &signature.InvalidSignatureError{Msg: fmt.Sprintf("these required headers are not marked as critical: %v", keys)}
+	}
+
+	return nil
+}
+
+func getSignatureAlgorithm(alg string) (signature.Algorithm, error) {
+	signatureAlg, ok := jwsAlgSignatureAlgMap[alg]
+	if !ok {
+		return 0, &signature.UnsupportedSignatureAlgoError{Alg: alg}
+	}
+
+	return signatureAlg, nil
+}
+
+func getExtendedAttributes(attrs map[string]interface{}, critical []string) []signature.Attribute {
+	extendedAttr := make([]signature.Attribute, 0, len(attrs))
+	for key, value := range attrs {
+		extendedAttr = append(extendedAttr, signature.Attribute{
+			Key:      key,
+			Critical: contains(critical, key),
+			Value:    value,
+		})
+	}
+	return extendedAttr
+}
+
+func contains(s []string, e string) bool {
+	for _, a := range s {
+		if a == e {
+			return true
+		}
+	}
+	return false
+}
+
+func generateJWS(compact string, req *signature.SignRequest, certs []*x509.Certificate) (*jwsEnvelope, error) {
+	parts := strings.Split(compact, ".")
+	if len(parts) != 3 {
+		// this should never happen
+		return nil, errors.New("unexpected error occurred while generating a JWS-JSON serialization from compact serialization")
+	}
+
+	rawCerts := make([][]byte, len(certs))
+	for i, cert := range certs {
+		rawCerts[i] = cert.Raw
+	}
+
+	return &jwsEnvelope{
+		Protected: parts[0],
+		Payload:   parts[1],
+		Signature: parts[2],
+		Header: jwsUnprotectedHeader{
+			CertChain:    rawCerts,
+			SigningAgent: req.SigningAgent,
+		},
+	}, nil
+}
+
+// getSignerAttributes merge extended signed attributes and protected header to be signed attributes.
+func getSignedAttributes(req *signature.SignRequest, algorithm string) (map[string]interface{}, error) {
+	extAttrs := make(map[string]interface{})
+	crit := []string{headerKeySigningScheme}
+
+	// write extended signed attributes to the extAttrs map
+	for _, elm := range req.ExtendedSignedAttributes {
+		extAttrs[elm.Key] = elm.Value
+		if elm.Critical {
+			crit = append(crit, elm.Key)
+		}
+	}
+
+	jwsProtectedHeader := jwsProtectedHeader{
+		Algorithm:     algorithm,
+		ContentType:   req.Payload.ContentType,
+		SigningScheme: req.SigningScheme,
+	}
+
+	switch req.SigningScheme {
+	case signature.SigningSchemeX509:
+		jwsProtectedHeader.SigningTime = &req.SigningTime
+	case signature.SigningSchemeX509SigningAuthority:
+		crit = append(crit, headerKeyAuthenticSigningTime)
+		jwsProtectedHeader.AuthenticSigningTime = &req.SigningTime
+	}
```

Comment:
same here. Needs to add a default to reject unsupported SigningScheme.

Reply:
Added a default case.